### DISCUSSION TITLE
Reducing Memory and CPU Usage w/ Pyright?

### DISCUSSION THREAD

**Discussion by jyue86** [microsoft/pyright#4941] [Q&A] [answered]
I've recently gotten started with nvim and using pyright as my Python LSP, and I've noticed that for larger-ish codebases, my CPU and RAM usage intensifies.

But I'm guessing at this time, I might not be configuring PyRight properly. Currently, I'm working in this [repository](https://github.com/openai/improved-diffusion). I open `improved_diffusion/image_datasets.py`, and immediately, my CPU usage increases to 50% and stayed that way until I closed nvim. The node process that belongs to Pyright eventually peaks at 2 GB. During this coding session, I didn't really do much because I quickly noticed that my auto-suggestions were being slow,. To experiment, I would type basic things  many times such as `print`, `if`, `import`, and other things in random places, so I could test how fast Pyright could auto-suggest things At first, the autosuggestions would appear albeit somewhat slow, but as the session went on, they would not appear at all. Later, I believe Pyright crashed, and I received the following error message in `LspLog`: 

```
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	"\n<--- Last few GCs --->\n\n[11254:0x6acd2d0]   506887 ms: Mark-sweep (reduce) 2034.3 (2083.1) -> 2033.4 (2082.6) MB, 1549.0 / 0.0 ms  (average mu = 0.102, current mu = 0.002) allocation failure; scavenge might not succeed\n[11254:0x6acd2d0]   507578 ms: Mark-sweep (reduce) 2034.8 (2082.9) -> 2033.6 (2082.9) MB, 688.2 / 0.0 ms  (average mu = 0.069, current mu = 0.004) allocation failure; scavenge might not succeed\n\n\n<--- JS stacktrace --->\n\nFATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	" 1: 0xb7b3e0 node::Abort() [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	" 2: 0xa8c8aa  [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	" 3: 0xd69100 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	" 4: 0xd694a7 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	" 5: 0xf46ba5  [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	" 6: 0xf47aa8 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	" 7: 0xf57fb3  [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	" 8: 0xf58e28 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	" 9: 0xf3378e v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	"10: 0xf34b57 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	"11: 0xf15d2a v8::internal::Factory::NewFillerObject(int, v8::internal::AllocationAlignment, v8::internal::AllocationType, v8::internal::AllocationOrigin) [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	"12: 0x12dacdf v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [node]\n"
[ERROR][2023-04-11 16:49:22] .../vim/lsp/rpc.lua:733	"rpc"	"pyright-langserver"	"stderr"	"13: 0x1707b79  [node]\n"
```
I believe this crash probably happened because of point 4 in this [comment](https://github.com/microsoft/pyright/issues/4900#issuecomment-1496849162)?

Here is my current configuration:
pyrightconfig.json:
```
{
  "include": [
    "."
  ],
  "exclude": [
    "**/__pycache__",
  ],
  "defineConstant": {
    "DEBUG": false
  },
  "venv": "venv",
  "reportMissingImports": true,
  "reportMissingTypeStubs": false,
  "pythonVersion": "3.10",
  "pythonPlatform": "Linux"
}
```

And my pyright is setup with the following options in my neovim config:
```
function()
require("pyright").setup({
	server = vim.tbl_extend("force", lsp_config, {
		on_attach = function(_, bufnr)
			on_attach(_, bufnr, PYTHON_SERVER)
		end,
		settings = {
			python = {
				analysis = {
					autoSearchPaths = true,
					useLibraryCodeForTypes = true,
					diagnosticMode = "openFilesOnly",
				},
			},
		},
	}),
})
end,

```
Other useful information:
pyright version - 1.1.302
node version - v18.15.0

What can I do to improve my current configuration and/or should I escalate this to a bug? Perhaps my issue will be fixed by the patch mentioned in the earlier issue that I linked to?

**Comment by erictraut** [comment]
I'm not able to repro the problem in VS Code using the same settings. I cloned the repo, followed the setup instructions, and opened the `improved_diffusion/image_datasets.py` file. Pyright took only a few dozen milliseconds to complete full analysis of the file.

I wasn't able to successfully install the mpi4py dependency; I'm not sure if that's related.

Additional log details may be helpful in diagnosing the problem. You can set "python.analysis.logLevel" to "Trace". This will cause more details to be logged.

We've seen situations where a virtual environment has been configured with a symlink to the root of the drive or the `/usr` directory, and that can cause performance issues because pyright will attempt to find all of the ".py" files on the drive. It may be worth creating a clean venv and trying to repro the problem using that.

**Reply by jyue86** [reply]
I did not have problems installing the mpi4py dependency, so I believe this is not the issue. 

I failed to mention earlier that that the codebase I'm working in is a slightly modified version of OpenAI's improved diffusion repository. Based on your response, I re-cloned a clean copy, and I am not running into the high cpu/ram usage issue. So I believe there must be something inside my  modified codebase that must be causing the issue, and perhaps including the paths inside my `pyrightconfig.json`'s `exclude` key could likely solve the issue?

I've set `python.analysis.logLevel` to "Trace", but I am unsure where the logs are placed. Where can I find them?

**Reply by erictraut** [reply]
I don't know where to find the logs when using neovim. In VS Code, they are in the "Output" panel. You might want to ask someone who is familiar with neovim's LSP support.

**Reply by erictraut** [reply]
Also, please try to repro the problem with pyright 1.1.303, which I published yesterday. It contains a fix for a bug that could explain what you're seeing.

**Reply by jyue86** [reply]
Unfortunately, I am still able to repro the problem with the latest update.

I have also found a way to view the logs in neovim. For future reference, it is indeed running `LspLog` in command mode, but before, `lua vim.lsp.set_log_level('trace')` must also be run in command mode.

Attached is a zipped folder that contains the log. There is unfortunately *a lot* of lines, and I had to close my neovim before even more lines could be added. Hopefully, the first few hundred lines will be able to give an idea of the problem, and I appreciate your help in solving this.


[pyright_log.zip](https://github.com/microsoft/pyright/files/11218243/pyright_log.zip)

**Reply by erictraut** [reply]
@jyue86, those are not the logs I'm looking for. Those are logs of the LSP messages being passed between the client and the language server. I'm interested in seeing the console logs.

You mentioned that you're working from a modified codebase. Is it checked in to a public fork, perhaps? I won't be able to diagnose this without a repro or logs.

Another option is for you to try to repro with the command-line version of pyright or the VS Code extension.

**Reply by jyue86** [reply]
Yes, the owner has allowed me to share the repository, [link here](https://github.com/Nanze-Chen/diffusion-debug). I also am not able to reproduce this on my VS Code, but on the command line, I run `pyright --verbose .` in the `diffusion-debug` repository. Unless the command line version is designed this way, pyright doesn't stop until I kill it, and I'm noticing similar behavior (high ram/ cpu usage).

[pyright_logs.txt](https://github.com/microsoft/pyright/files/11226370/pyright_logs.txt)

Also, if you're inclined to try reproducing with neovim, you can find my neovim configuration in my [Dotfiles repository](https://github.com/jyue86/DotFiles).

**Reply by erictraut** [reply, marked as answer]
Based on the logs, it appears what's happening is that you've created a virtual environment called `venv` in your project directory, and pyright is treating the entire contents of this venv as part of your project. The logs indicate that pyright found 11727 source files.

Pyright attempts to detect and exclude any directory that appears to be a virtual environment, but it applies these "auto excludes" only if your configuration file doesn't provide explicit excludes. Since your configuration file supplies an "exclude" section, these "auto excludes" are not applied. Pyright's auto-exclude behavior also excludes `**/__pycache__`, so there's no need to explicitly exclude this. If you do supply an "exclude" section, you need to ensure that you've excluded everything including `venv`.

**Reply by jyue86** [reply]
Yes, this seems to have fixed the issue. Thank you very much!
